### ISSUE TITLE
Panic in `validate_classinfo_name`

### ISSUE THREAD

**Issue by schubart** [staktrace/cafebabe#7]
## Background

I was experimenting with [AFL fuzzing](https://rust-fuzz.github.io/book/afl.html) and it triggered a panic. I think `cafebabe::parse_class` should never panic. If anything is wrong with the class file, an error should be returned.

## Details

`validate` does this:

``` rust
ConstantPoolEntry::ClassInfo(x) => Ok(
             x.ensure_type(ConstantPoolEntryTypes::UTF8)? && 
             x.borrow().get().validate_classinfo_name()?),
```

`ensure_type(UTF8)` returns true for `ConstantPoolEntry::Utf8` and for `ConstantPoolEntry::Utf8Bytes`. So `validate_classinfo_name` might get called with either of these two variants. It then does this:

``` rust
    fn validate_classinfo_name(&self) -> Result<bool, ParseError> {
        match self {
            ConstantPoolEntry::Utf8(x) => { ... }
            _ => panic!("Attempting to get utf-8 data from non-utf8 constant pool entry!"),
        }
    }
```

So it panics if called with `Utf8Bytes`. The same issue exists for some other `validate_*` functions.

I see two issues here:

## Constant pool entry type checks are out of sync

The panic happens because `validate` and `validate_classinfo_name` both check if the constant pool entry has the correct type, but they have different ideas about what the allowed types are.

I think this could be fixed by making `validate_classinfo_name` cal `fail!` instead of `panic!` and by removing the call to `ensure_type` in `validate`. Perhaps this could get rid of `ensure_type` and even `ConstantPoolEntryTypes` entirely. This would seem to make sense because Rust already knows at runtime what variant of `ConstantPoolEntry` it is, so it seems awkward to have a parallel `ConstantPoolEntryTypes` that mirrors `ConstantPoolEntry`.

## Better to fail if `cesu8::from_java_cesu8` fails?

I understand that "UTF-8" constant pool entries in Java class files are not standard UTF-8 [(spec)](https://docs.oracle.com/javase/specs/jvms/se17/html/jvms-4.html#jvms-4.4.7). But isn't that handled by the `cesu8` crate? If even that method cannot make sense of the bytes, is it really worth keeping parsing? Holding on to these malformed bytes seems to be responsible for awkward code in several places and ultimately, these bytes will be passed to the end user (using `StringBytes`), who also won't know how to handle them.

I think it would make more sense to `fail!` in this case. However, I see that this is how the library used to behave until you changed it [here](https://github.com/staktrace/cafebabe/commit/8a24f70900143d14c2c37a9e58cee2c7e39b4815#diff-633e2e5ab5d3fabbab5c270cb244aae639580b0c997370422bbef33866f248ab). I'm curious why you made that change? Did you encounter such badly formed strings "in the wild"?

**Comment by staktrace** [comment]
I'm busy this weekend so I'll look into this more early next week, but in response to your last question, yes, I believe I ran into this case in the JDK 16 standard library (the one that gets parsed with `cargo test`). I think there is still value in keeping that because the library shouldn't discard information that some consumers of the library might need.

**Comment by schubart** [comment]
You're right, such odd bytes are found in `openjdk16.0.1-modules/jdk.charsets/sun/nio/cs/ext/MS932.class`. I should have checked that before asking the question.

**Comment by schubart** [comment]
FYI, this is how a similar library deals with it:
https://github.com/Palmr/classfile-parser/pull/21
